be/src/exprs/string_functions.cpp

```diff
@@ -535,17 +526,47 @@ StringVal StringFunctions::concat_ws(
     uint8_t* ptr = result.ptr;
 
     // Loop again to append the data.
-    memcpy(ptr, strs[0].ptr, strs[0].len);
-    ptr += strs[0].len;
-    for (int32_t i = 1; i < num_children; ++i) {
-        memcpy(ptr, sep.ptr, sep.len);
-        ptr += sep.len;
+    for (int32_t i = 0; i < num_children; ++i) {
         memcpy(ptr, strs[i].ptr, strs[i].len);
         ptr += strs[i].len;
     }
     return result;
 }
 
+StringVal StringFunctions::concat_ws(
+        FunctionContext* context, const StringVal& sep, 
+        int num_children, const StringVal* strs) {
+    DCHECK_GE(num_children, 1);
+    if (sep.is_null) {
+        return StringVal::null();
+    }
+
+    int32_t total_size = strs[0].is_null ? 0 : strs[0].len;
+    // Loop once to compute the final size and reserve space.
+    for (int32_t i = 1; i < num_children; ++i) {
+        total_size += strs[i].is_null ? 0 : (sep.len + strs[i].len);
```

Comment:
if strs[0] is null, total_size would add a more `sep.len`.

same with your data copy logical

Reply:
computing size should be equal with appending data